includes/ziti/ziti_buffer.h

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright (c) 2021 NetFoundry Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef ZITI_SDK_ZITI_BUFFER_H
+#define ZITI_SDK_ZITI_BUFFER_H
+
+#include <stdint.h>
+#include "externs.h"
+#include "ziti_log.h"
+
+#if !defined(__DEFINED_ssize_t) && !defined(__ssize_t_defined)
+#if _WIN32
+typedef intptr_t ssize_t;
+#define __DEFINED_ssize_t
+#define __ssize_t_defined
+#else
+#include <unistd.h>
+#endif
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Growing string buffer.
+ *
+ * string_buf allocates memory internally as needed unless it was created with [new_fixed_string_buf()]
+ */
+typedef struct string_buf_s string_buf_t;
+
+/**
+ * Create new string buffer.
+ * @return new buffer instance
+ */
+ZITI_FUNC string_buf_t *new_string_buf();
+
+/**
+ * Create buffer using passed in memory for output. No new memory will be allocated.
+ * @param outbuf
+ * @param max
+ * @return new buffer instance
+ */
+ZITI_FUNC string_buf_t *new_fixed_string_buf(char *outbuf, size_t max);
+
+/**
+ * Deallocate all memory associated with the given string buffer.
+ *
+ * The passed in pointer is not valid after the function invocation.
+ * @param wb
+ */
+ZITI_FUNC void delete_string_buf(string_buf_t *wb);
+
+/**
+ * Append `\0` terminated string to the buffer.
```

Comment:
as an API reader, I wonder, what would happen if i were to pass a string which is not null terminated. would you append it for me or will it end up allowing me to to read past the buffer after i append a non-terminated string?